### DISCUSSION TITLE
Nodejs- failed to launch: exec.d: failed to execute exec.d file at path '/layers/paketo-buildpacks_npm-install/launch-modules/exec.d/0-setup-symlinks': exit status 1

### DISCUSSION THREAD

**Discussion by skagithala-clgx** [paketo-buildpacks/feedback#325] [Node.js Team]
Hi Team,

when we run nodejs applcation with pack build, build is showing as successful, while deploying to kubernetes pod received below error
could you please let me know any parameter needs to add to the pack command  
ERROR: failed to launch: exec.d: failed to execute exec.d file at path '/layers/paketo-buildpacks_npm-install/launch-modules/exec.d/0-setup-symlinks': exit status 1 

pipeline log:

03:35:05  ===> DETECTING
03:35:06  [detector] target distro name/version labels not found, reading /etc/os-release file
03:35:06  [detector] 6 of 10 buildpacks participating
03:35:06  [detector] paketo-buildpacks/ca-certificates 3.8.6
03:35:06  [detector] paketo-buildpacks/node-engine     5.0.4
03:35:06  [detector] paketo-buildpacks/npm-install     1.6.4
03:35:06  [detector] paketo-buildpacks/node-run-script 1.0.36
03:35:06  [detector] paketo-buildpacks/node-start      2.1.11
03:35:06  [detector] paketo-buildpacks/npm-start       2.0.20
03:35:06  ===> RESTORING
03:35:07  ===> BUILDING
03:35:07  [builder] target distro name/version labels not found, reading /etc/os-release file
03:35:07  [builder] 
03:35:07  [builder] Paketo Buildpack for CA Certificates 3.8.6
03:35:07  [builder]   https://github.com/paketo-buildpacks/ca-certificates
03:35:07  [builder]   Build Configuration:
03:35:07  [builder]     $BP_EMBED_CERTS                    false  Embed certificates into the image
03:35:07  [builder]     $BP_ENABLE_RUNTIME_CERT_BINDING    true   Deprecated: Enable/disable certificate helper layer to add certs at runtime
03:35:07  [builder]     $BP_RUNTIME_CERT_BINDING_DISABLED  false  Disable certificate helper layer to add certs at runtime
03:35:07  [builder]   Launch Helper: Contributing to layer
03:35:07  [builder]     Creating /layers/paketo-buildpacks_ca-certificates/helper/exec.d/ca-certificates-helper
03:35:07  [builder] Paketo Buildpack for Node Engine 5.0.4
03:35:07  [builder]   Resolving Node Engine version
03:35:07  [builder]     Candidate version sources (in priority order):
03:35:07  [builder]                 -> ""
03:35:07  [builder]       <unknown> -> ""
03:35:07  [builder] 
03:35:07  [builder]     Selected Node Engine version (using ): 20.18.0
03:35:07  [builder] 
03:35:07  [builder]   Executing build process
03:35:07  [builder]     Installing Node Engine 20.18.0
03:35:14  [builder]       Completed in 6.665s
03:35:14  [builder] 
03:35:14  [builder]   Generating SBOM for /layers/paketo-buildpacks_node-engine/node
03:35:14  [builder]       Completed in 0s
03:35:14  [builder] 
03:35:14  [builder]   Configuring build environment
03:35:14  [builder]     NODE_ENV     -> "production"
03:35:14  [builder]     NODE_HOME    -> "/layers/paketo-buildpacks_node-engine/node"
03:35:14  [builder]     NODE_OPTIONS -> "--use-openssl-ca"
03:35:14  [builder]     NODE_VERBOSE -> "false"
03:35:14  [builder] 
03:35:14  [builder]   Configuring launch environment
03:35:14  [builder]     NODE_ENV     -> "production"
03:35:14  [builder]     NODE_HOME    -> "/layers/paketo-buildpacks_node-engine/node"
03:35:14  [builder]     NODE_OPTIONS -> "--use-openssl-ca"
03:35:14  [builder]     NODE_VERBOSE -> "false"
03:35:14  [builder] 
03:35:14  [builder]     Writing exec.d/0-optimize-memory
03:35:14  [builder]       Calculates available memory based on container limits at launch time.
03:35:14  [builder]       Made available in the MEMORY_AVAILABLE environment variable.
03:35:14  [builder]     Writing exec.d/1-inspector
03:35:14  [builder] 
03:35:14  [builder] Paketo Buildpack for NPM Install 1.6.4
03:35:14  [builder]   Resolving installation process
03:35:14  [builder]     Process inputs:
03:35:14  [builder]       node_modules      -> "Not found"
03:35:14  [builder]       npm-cache         -> "Not found"
03:35:14  [builder]       package-lock.json -> "Not found"
03:35:14  [builder] 
03:35:14  [builder]     Selected NPM build process: 'npm install'
03:35:14  [builder] 
03:35:14  [builder]   Executing build environment install process
03:35:14  [builder]     Running 'npm install --unsafe-perm --cache /layers/paketo-buildpacks_npm-install/npm-cache'
03:35:29  [builder]       
03:35:29  [builder]       added 406 packages, and audited 407 packages in 14s
03:35:29  [builder]       
03:35:29  [builder]       36 packages are looking for funding
03:35:29  [builder]         run `npm fund` for details
03:35:29  [builder]       
03:35:29  [builder]       5 vulnerabilities (2 moderate, 1 high, 2 critical)
03:35:29  [builder]       
03:35:29  [builder]       To address all issues possible (including breaking changes), run:
03:35:29  [builder]         npm audit fix --force
03:35:29  [builder]       
03:35:29  [builder]       Some issues need review, and may require choosing
03:35:29  [builder]       a different dependency.
03:35:29  [builder]       
03:35:29  [builder]       Run `npm audit` for details.
03:35:29  [builder]       npm notice
03:35:29  [builder]       npm notice New minor version of npm available! 10.8.2 -> 10.9.2
03:35:29  [builder]       npm notice Changelog: https://github.com/npm/cli/releases/tag/v10.9.2
03:35:29  [builder]       npm notice To update run: npm install -g npm@10.9.2
03:35:29  [builder]       npm notice
03:35:30  [builder]       Completed in 15.812s
03:35:30  [builder] 
03:35:30  [builder]   Configuring build environment
03:35:30  [builder]     NODE_ENV -> "development"
03:35:30  [builder]     PATH     -> "/layers/paketo-buildpacks_npm-install/build-modules/node_modules/.bin_local:$PATH:/layers/paketo-buildpacks_npm-install/build-modules/node_modules/.bin"
03:35:30  [builder] 
03:35:30  [builder]   Generating SBOM for /layers/paketo-buildpacks_npm-install/build-modules
03:35:31  [builder]       Completed in 1.423s
03:35:31  [builder] 
03:35:31  [builder]   Executing launch environment install process
03:35:32  [builder]     Running 'npm prune'
03:35:34  [builder]       
03:35:34  [builder]       up to date, audited 407 packages in 1s
03:35:34  [builder]       
03:35:34  [builder]       36 packages are looking for funding
03:35:34  [builder]         run `npm fund` for details
03:35:34  [builder]       
03:35:34  [builder]       5 vulnerabilities (2 moderate, 1 high, 2 critical)
03:35:34  [builder]       
03:35:34  [builder]       To address all issues possible (including breaking changes), run:
03:35:34  [builder]         npm audit fix --force
03:35:34  [builder]       
03:35:34  [builder]       Some issues need review, and may require choosing
03:35:34  [builder]       a different dependency.
03:35:34  [builder]       
03:35:34  [builder]       Run `npm audit` for details.
03:35:34  [builder]       npm notice
03:35:34  [builder]       npm notice New minor version of npm available! 10.8.2 -> 10.9.2
03:35:34  [builder]       npm notice Changelog: https://github.com/npm/cli/releases/tag/v10.9.2
03:35:34  [builder]       npm notice To update run: npm install -g npm@10.9.2
03:35:34  [builder]       npm notice
03:35:35  [builder]       Completed in 1.276s
03:35:35  [builder] 
03:35:35  [builder]   Configuring launch environment
03:35:35  [builder]     NODE_PROJECT_PATH   -> "/workspace"
03:35:35  [builder]     NPM_CONFIG_LOGLEVEL -> "error"
03:35:35  [builder]     PATH                -> "/layers/paketo-buildpacks_npm-install/launch-modules/node_modules/.bin_local:$PATH:/layers/paketo-buildpacks_npm-install/launch-modules/node_modules/.bin"
03:35:35  [builder] 
03:35:35  [builder]   Generating SBOM for /layers/paketo-buildpacks_npm-install/launch-modules
03:35:36  [builder]       Completed in 1.254s
03:35:36  [builder] 
03:35:36  [builder] 
03:35:37  [builder] Paketo Buildpack for Node Run Script 1.0.36
03:35:37  [builder]   Executing build process
03:35:37  [builder]     Running 'npm run ci_build'
03:35:37  [builder]       
03:35:37  [builder]       > tax-us-researchhub-tenantproxy@9.0.22 ci_build
03:35:37  [builder]       > rm -rf config && mv config_saas config  && npm install
03:35:37  [builder]       
03:35:37  [builder]       npm warn reify Removing non-directory /workspace/node_modules
03:35:40  [builder]       npm warn deprecated glob@7.1.6: Glob versions prior to v9 are no longer supported
03:35:40  [builder]       npm warn deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
03:35:40  [builder]       npm warn deprecated npmlog@5.0.1: This package is no longer supported.
03:35:41  [builder]       npm warn deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
03:35:41  [builder]       npm warn deprecated har-validator@5.1.5: this library is no longer supported
03:35:41  [builder]       npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
03:35:41  [builder]       npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
03:35:41  [builder]       npm warn deprecated crypto@1.0.1: This package is no longer supported. It's now a built-in Node module. If you've depended on crypto, you should switch to the one that's built-in.
03:35:41  [builder]       npm warn deprecated gauge@3.0.2: This package is no longer supported.
03:35:41  [builder]       npm warn deprecated are-we-there-yet@2.0.0: This package is no longer supported.
03:35:41  [builder]       npm warn deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
03:35:45  [builder]       
03:35:45  [builder]       added 406 packages, and audited 407 packages in 8s
03:35:45  [builder]       
03:35:45  [builder]       36 packages are looking for funding
03:35:45  [builder]         run `npm fund` for details
03:35:45  [builder]       
03:35:45  [builder]       5 vulnerabilities (2 moderate, 1 high, 2 critical)
03:35:45  [builder]       
03:35:45  [builder]       To address all issues possible (including breaking changes), run:
03:35:45  [builder]         npm audit fix --force
03:35:45  [builder]       
03:35:45  [builder]       Some issues need review, and may require choosing
03:35:45  [builder]       a different dependency.
03:35:45  [builder]       
03:35:45  [builder]       Run `npm audit` for details.
03:35:45  [builder] 
03:35:45  [builder]     Running 'npm run upload'
03:35:45  [builder]       
03:35:45  [builder]       > application@9.0.22 upload
03:35:45  [builder]       > npm publish --userconfig tax-us-researchhub-tenantproxy.npmrc
03:35:45  [builder]       
03:35:46  [builder]       npm notice
03:35:46  [builder]       npm notice package: tax-us-researchhub-tenantproxy@9.0.22
03:35:46  [builder]       npm notice Tarball Contents
03:35:46  [builder]       npm notice 144B CODEOWNERS
03:35:46  [builder]       npm notice 4.6kB Jenkinsfile
03:35:46  [builder]       npm notice 777B README.md
03:35:46  [builder]       npm notice 1.3kB common/encryption.js
03:35:46  [builder]       npm notice 1.1kB common/error.js
03:35:46  [builder]       npm notice 1.0kB common/jwt.js
03:35:46  [builder]       npm notice 2.0kB config_saas_local/config.js
03:35:46  [builder]       npm notice 56B config_saas_local/serviceurls.js
03:35:46  [builder]       npm notice 3.1kB config/config.js
03:35:46  [builder]       npm notice 2.0kB config/data.js
03:35:46  [builder]       npm notice 56B config/serviceurls.js
03:35:46  [builder]       npm notice 1.6kB package.json
03:35:46  [builder]       npm notice 5.3kB routes.js
03:35:46  [builder]       npm notice 4.1kB routes/adminauth.js
03:35:46  [builder]       npm notice 3.3kB routes/tenantproxy.js
03:35:46  [builder]       npm notice 1.5kB routes/validation/inputvalidator.js
03:35:46  [builder]       npm notice 0B routes/validation/schema.js
03:35:46  [builder]       npm notice 3.3kB server.js
03:35:46  [builder]       npm notice 6.6kB serviceregistry/registry.js
03:35:46  [builder]       npm notice 4.9kB swaggerdocs/model-schema.yaml
03:35:46  [builder]       npm notice 1.4kB swaggerdocs/swagger.js
03:35:46  [builder]       npm notice 335B tax-us-researchhub-tenantproxy.npmrc
03:35:46  [builder]       npm notice 16B test.txt
03:35:46  [builder]       npm notice 1.5kB tracer/shim.js
03:35:46  [builder]       npm notice 5.8kB tracer/tracer.js
03:35:46  [builder]       npm notice Tarball Details
03:35:46  [builder]       npm notice name: tax-us-researchhub-tenantproxy
03:35:46  [builder]       npm notice version: 9.0.22
03:35:46  [builder]       npm notice filename: tax-us-researchhub-tenantproxy-9.0.22.tgz
03:35:46  [builder]       npm notice package size: 14.1 kB
03:35:46  [builder]       npm notice unpacked size: 55.6 kB
03:35:46  [builder]       npm notice shasum: 5963f22378d93bd0985171a8e4d77dc4bb95c340
03:35:46  [builder]       npm notice integrity: sha512-Im1dNL2APDSkX[...]YEB1jpLJkbdQg==
03:35:46  [builder]       npm notice total files: 25
03:35:46  [builder]       npm notice
03:35:46  [builder]       npm notice Publishing to https://artifactory/npm-release-virtual/ with tag latest and default access
03:35:46  [builder]       + application@9.0.22
03:35:46  [builder] 
03:35:46  [builder]     Completed in 9.189s
03:35:46  [builder] 
03:35:46  [builder] Paketo Buildpack for Node Start 2.1.11
03:35:46  [builder]   Assigning launch processes:
03:35:46  [builder]     web (default): node server.js
03:35:46  [builder] 
03:35:46  [builder] Paketo Buildpack for NPM Start 2.0.20
03:35:46  [builder]   Assigning launch processes:
03:35:46  [builder]     web (default): sh /workspace/start.sh
03:35:46  [builder] 
03:35:47  ===> EXPORTING
03:35:48  [exporter] Adding layer 'paketo-buildpacks/ca-certificates:helper'
03:35:49  [exporter] Adding layer 'paketo-buildpacks/node-engine:node'
03:35:54  [exporter] Adding layer 'paketo-buildpacks/npm-install:launch-modules'
03:35:56  [exporter] Adding layer 'buildpacksio/lifecycle:launch.sbom'
03:35:59  [exporter] Added 1/1 app layer(s)
03:35:59  [exporter] Adding layer 'buildpacksio/lifecycle:launcher'
03:35:59  [exporter] Adding layer 'buildpacksio/lifecycle:config'
03:35:59  [exporter] Adding layer 'buildpacksio/lifecycle:process-types'
03:35:59  [exporter] Adding label 'io.buildpacks.lifecycle.metadata'
03:35:59  [exporter] Adding label 'io.buildpacks.build.metadata'
03:35:59  [exporter] Adding label 'io.buildpacks.project.metadata'
03:35:59  [exporter] Setting default process type 'web'

**Comment by pacostas** [comment]
Can you please provide instructions on how to reproduce it?

**Reply by skagithala-clgx** [reply]
We tested the pipeline with internal application. Tried below command to build the application 

pack build us-docker.pkg.dev/GAR path --builder paketobuildpacks/builder-jammy-full -e 'NODE_ENV=development' -e 'BP_NODE_RUN_SCRIPTS=ci_build,upload'

When we tried with below packeto supplied repo didn't find any issue, deployed successfully.
https://github.com/paketo-buildpacks/samples/tree/main/nodejs/npm

command executed:
pack build us-docker.pkg.dev/GAR path --builder us-docker.pkg.dev/paketobuildpacks/builder-jammy-base -e NODE_ENV=development -e BP_NODE_RUN_SCRIPTS=upload
please let me know if anything required from my side

**Reply by pacostas** [reply]
Yes, indeed it builds fine. Do you have a command that reproduces the error while deploying on kubernetes? I don't have any other way to test it on my pc

**Reply by skagithala-clgx** [reply]
we have deploying through helm charts, below is the command 
pipeline output
+ helm upgrade --install --wait --timeout 15m application-name
 /home/jenkins/agent/workspace/earchhub-tenantproxy_feature_dev/HELM_DIR/deployment/ -f /home/jenkins/agent/workspace/earchhub-tenantproxy_feature_dev/HELM_DIR/deployment/defaultValues.yaml -f override-values.yaml
03:38:27  Release "application-name" does not exist. Installing it now.
03:53:34  Error: context deadline exceeded

NAME                                                               READY   STATUS             RESTARTS        AGE
pod/application-name                                               1/2     CrashLoopBackOff   35 (78s ago)    155m

PS C:\Users\Srikanth_Kagithala> kubectl logs pod/application-name
2024/12/18 15:22:19 readlink /workspace/node_modules: invalid argument
ERROR: failed to launch: exec.d: failed to execute exec.d file at path '/layers/paketo-buildpacks_npm-install/launch-modules/exec.d/0-setup-symlinks': exit status 1

**Reply by skagithala-clgx** [reply]
did you get chance to replicate the issue

**Reply by pacostas** [reply]
Sorry but I dont have enough context to replicate it, my wild guess is that you build the app in a different environment than you run it in terms of architecture
